### PR TITLE
[WIP] fix  #6999 (allow [nimDocOnly] in commit msg);  fix #13477 (flaky nodejs install); log PR url (+other info) in runCI logs

### LINKED ISSUE #13477: [CI] `Install node.js 8.x` flaky step
I've encountered this a few times already
example:
https://dev.azure.com/nim-lang/Nim/_build/results?buildId=2806&view=logs&jobId=a0440cd6-2060-5545-8b53-639e777de0c6&j=a0440cd6-2060-5545-8b53-639e777de0c6&t=3a9621cd-7d4e-5bb0-bcbd-e24390a180e8

### Example
* for `packages OSX_amd64_cpp` in https://github.com/nim-lang/Nim/pull/13305
* and then after restarting CI happened again but this time for Windows_amd64_pkg: https://dev.azure.com/nim-lang/Nim/_build/results?buildId=2807&view=logs&jobId=f6c2f757-7152-5b00-b42b-aaebf1a063dd&j=f6c2f757-7152-5b00-b42b-aaebf1a063dd&t=6c4f6d65-82fc-5149-e4b2-e3852d9699f3

```
2020-02-24T03:14:06.1047600Z ##[section]Starting: Install node.js 8.x
2020-02-24T03:14:06.1057800Z ==============================================================================
2020-02-24T03:14:06.1058270Z Task         : Node.js tool installer
2020-02-24T03:14:06.1058730Z Description  : Finds or downloads and caches the specified version spec of Node.js and adds it to the PATH
2020-02-24T03:14:06.1059170Z Version      : 0.160.0
2020-02-24T03:14:06.1059460Z Author       : Microsoft Corporation
2020-02-24T03:14:06.1059880Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/tool/node-js
2020-02-24T03:14:06.1060350Z ==============================================================================
2020-02-24T03:14:06.3224040Z (node:816) Warning: Use Cipheriv for counter mode of aes-256-ctr
2020-02-24T03:14:06.3225120Z (node:816) Warning: Use Cipheriv for counter mode of aes-256-ctr
2020-02-24T03:14:06.3225980Z (node:816) Warning: Use Cipheriv for counter mode of aes-256-ctr
2020-02-24T03:14:06.3226730Z (node:816) Warning: Use Cipheriv for counter mode of aes-256-ctr
2020-02-24T03:14:06.3227830Z (node:816) Warning: Use Cipheriv for counter mode of aes-256-ctr
2020-02-24T03:14:06.3229220Z (node:816) Warning: Use Cipheriv for counter mode of aes-256-ctr
2020-02-24T03:14:06.4237670Z Downloading: https://nodejs.org/dist/v8.17.0/node-v8.17.0-darwin-x64.tar.gz
2020-02-24T03:15:19.1448340Z Extracting archive
2020-02-24T03:15:19.1701960Z [command]/usr/bin/tar xzC /Users/runner/runners/2.165.0/work/_temp/6966d1b1-6cd2-4d50-af37-1ea80e85db3a -f /Users/runner/runners/2.165.0/work/_temp/b01f0c58-f3d2-4011-95b8-13349707640d
2020-02-24T03:15:19.2960400Z node-v8.17.0-darwin-x64/lib/node_modules/npm/node_modules/uid-number/uid-number.js: (Empty error message)
2020-02-24T03:15:19.3086320Z tar: Error exit delayed from previous errors.
2020-02-24T03:15:19.3216120Z ##[error]/usr/bin/tar failed with return code: 1
2020-02-24T03:15:19.3233510Z ##[section]Finishing: Install node.js 8.x
```

### PR DESCRIPTION
**Pull request by timotheecour**
EDIT 2020-05-13: this PR is useful but needs to be updated to reflect recent changes in how CI works; in particular can be modified to handle `[ci skip]` which isn't working in azure pipelines)

* fix https://github.com/nim-lang/Nim/pull/6999 as follows: if the **last commit msg** containing `[nimDocOnly]` will only build nim and run nim doc-related stuff
* this results in 7.5X speedups for `[nimDocOnly]`commits:  ~ 4mn per machine instead of ~30mn 
* fix #13477 `Install node.js 8.x` flaky step
#13477 keeps happening regularly (and even breaks with different error messages depending on the time) ; and it's simpler anyway to install directly from system package installers, eg `brew` on OSX ; furthermore, moving more logic into Nim code (`installNode `) is better (less dependency on CI specific logic, which makes it easier to switch as needed or work with multiple CI's)
* log a few other relevant variables that help when looking at runCI logs (the main log that contains almost all of nim CI), so they're self contained when shared
  * the PR url so it's easy to find originating PR when looking at those
  * the number of CPU's (helps when interpreting differences in timing results; some have 2, some have 4 in azure IIRC)
  * build ID + 1 or 2 more relevant info

eg for this PR, CI logs for the runCI step (eg like https://dev.azure.com/nim-lang/Nim/_build/results?buildId=2984&view=logs&j=a0440cd6-2060-5545-8b53-639e777de0c6)
will show this (and urls' are clickable):
```
...
hostOS: linux, hostCPU: amd64, nproc: 2, int: 8, float: 8, cpuEndian: littleEndian, cwd: /home/vsts/work/1/s
urlPR: https://github.com/nim-lang/Nim/pull/13556
urlCommit: https://github.com/nim-lang/Nim/commit/196c5b57f8e647a4219ba8f548fe8d4c43aaa608
branch: merge, mode: c, buildNum: 20200304.22
msg: 16c26ae4cb8f95d5694ccb57d6f3e407307e2f3d rerun without skipping tests
...
```
  * note this also works for CI logs triggered by a direct push to a branch as opposed to a PR (in which case the commit msg + url will be reflected as well, and isPR:false)

[EDIT]: that issue just happened again in https://dev.azure.com/nim-lang/Nim/_build/results?buildId=3007&view=logs&jobId=30931762-47c4-53b3-6a83-316eb5a6b9d7&j=30931762-47c4-53b3-6a83-316eb5a6b9d7&t=a385ca32-7847-5843-f526-f36e9a71a329

https://dev.azure.com/nim-lang/Nim/_build/results?buildId=3523&view=logs&jobId=1effb61f-3820-5ca5-606b-caa183b197b5&j=1effb61f-3820-5ca5-606b-caa183b197b5&t=17e5f679-7cb9-5688-0d47-09c257156869

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -152,4 +156,13 @@ jobs:
...
+        # see https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml
+        # most useful entries among the numerous defined variables
         SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+        NIM_CI_System_PullRequest_PullRequestNumber: $(System.PullRequest.PullRequestNumber)
+        NIM_CI_Build_SourceBranchName: $(Build.SourceBranchName)
+        NIM_CI_Build_Reason: $(Build.Reason)
+        NIM_CI_Build_Repository_Uri: $(Build.Repository.Uri)
+        NIM_CI_Build_SourceVersion: $(Build.SourceVersion)
+        NIM_CI_Build_SourceVersionMessage: $(Build.SourceVersionMessage)
+        NIM_CI_Build_BuildNumber: $(Build.BuildNumber)
```
Comment (alaviss) [inline azure-pipelines.yml]: Unlike SYSTEM_ACCESSTOKEN, all of these are exported by the CI agent. Just uppercase everything then replace `.` with `_`

File: tools/continuous_integration.nim [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import os, osproc, strformat
+
+proc isAzureCI*(): bool =
+  getEnv("NIM_CI_Build_SourceBranchName").len > 0
```
Comment (alaviss) [inline tools/continuous_integration.nim]: Please see how this is detected in testament.

**Review by alaviss — COMMENTED** [review]
> * log a few other relevant variables that help when looking at runCI logs (the main log that contains almost all of nim CI), so they're self contained when shared
>   
>   * the PR url so it's easy to find originating PR when looking at those
>   * the number of CPU's (helps when interpreting differences in timing results; some have 2, some have 4 in azure IIRC)
>   * build ID + 1 or 2 more relevant info

The idea isn't bad, but please actually format the data so that human can parse it rather than dump it all out.

File: tools/continuous_integration.nim (hunk 2 above)
Comment (timotheecour) [reply tools/continuous_integration.nim]: done, reused their code

File: azure-pipelines.yml (hunk 1 above)
Comment (timotheecour) [reply azure-pipelines.yml]: thx, done, reused existing code for that

**Comment by timotheecour** [comment]
> The idea isn't bad, but please actually format the data so that human can parse it rather than dump it all out.

done, shows:
```
hostOS: linux, hostCPU: amd64, nproc: 2, int: 8, float: 8, cpuEndian: littleEndian, cwd: /home/vsts/work/1/s
urlPR: https://github.com/nim-lang/Nim/pull/13556
urlCommit: https://github.com/nim-lang/Nim/commit/6124963f0e51edcb86b6f03fcfd250c08ede5ab6
branch: merge, mode: c, buildNum: 20200304.17
msg: 'Merge dc2d87b12dc96dd81bb4d79c6f66c16979b76f4f into a0eca7518223a18f3633150de2c8d3c1c9e71560'
```

**Pull request closed without merging**